oauthenticator/oauth2.py

```diff
@@ -426,8 +481,274 @@ def get_handlers(self, app):
+    def build_userdata_request_headers(self, access_token, token_type):
+        """
+        Builds and returns the headers to be used in the userdata request.
+        Called by the :meth:`oauthenticator.OAuthenticator.token_to_user`
+        """
+        return {
+            "Accept": "application/json",
+            "Content-Type": "application/json",
+            "User-Agent": "JupyterHub",
+            "Authorization": "{} {}".format(token_type, access_token),
+        }
+
+    def build_token_info_req_headers(self):
+        """
+        Builds and returns the headers to be used in the access token request.
+        Called by the :meth:`oauthenticator.OAuthenticator.get_tokens_info`.
+        """
+        headers = {"Accept": "application/json", "User-Agent": "JupyterHub"}
+
+        if self.basic_auth:
+            b64key = base64.b64encode(
+                bytes("{}:{}".format(self.client_id, self.client_secret), "utf8")
+            )
+            headers.update({"Authorization": "Basic {}".format(b64key.decode("utf8"))})
+        return headers
+
+    def user_info_to_username(self, user_info):
+        """
+        Gets the self.username_claim key's value from the user_info dictionary.
+        This is equivalent to the JupyterHub username.
+
+        Should be overridden by the authenticators for which the hub username cannot
+        be extracted this way and needs extra processing.
+
+        Args:
+            user_info: the dictionary returned by the userdata request
+
+        Returns:
+            user_info["self.username_claim"] or raises an error if such value isn't found.
+
+        Called by the :meth:`oauthenticator.OAuthenticator.authenticate`
+        """
+        username = user_info.get(self.username_claim, None)
+        if not username:
+            message = "No %s found in %s", self.username_claim, user_info
+            self.log.error(message)
+            raise ValueError(message)
+
+        return username
+
+    # Originally a GoogleOAuthenticator only feature
+    async def get_prev_refresh_token(self, handler, username):
+        """
+        Retrieves the `refresh_token` from previous encrypted auth state.
+        Called by the :meth:`oauthenticator.OAuthenticator.authenticate`
+        """
+        user = handler.find_user(username)
+        if not user or not user.encrypted_auth_state:
+            return
+
+        self.log.debug(
+            "Encrypted_auth_state was found, will try to decrypt and pull refresh_token from it..."
+        )
+
+        try:
+            encrypted = user.encrypted_auth_state
+            auth_state = await decrypt(encrypted)
+
+            return auth_state.get('refresh_token')
+        except (ValueError, InvalidToken, EncryptionUnavailable) as e:
+            self.log.warning(
+                f"Failed to retrieve encrypted auth_state for {username}. Error was {e}.",
+            )
+            return
+
+    def build_access_tokens_req_params(self, handler):
```

Comment:
Same as above, consistency:

```suggestion
    def build_access_tokens_request_params(self, handler):
```